api/bucketsd/service.go

```diff
@@ -970,16 +973,175 @@ func (s *Service) SetPath(ctx context.Context, req *pb.SetPathRequest) (res *pb.
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = s.setPath(ctx, &pb.SetPathRequest{
+		Key:  req.Key,
+		Path: cleanDst,
+		Cid:  n.Cid().String(),
+	}, dbID, dbToken, cleanDst)
+	if err != nil {
+		return nil, fmt.Errorf("error setting path: %v", err)
+	}
+	if buck.IsPrivate() {
+		meta, _, ok := buck.GetMetadataForPath(cleanPth, false)
+		if !ok {
+			return nil, fmt.Errorf("error getting metadata")
+		}
+		buck.SetMetadataAtPath(cleanDst, meta)
+		if err = s.Buckets.Verify(ctx, dbID, buck, tdb.WithToken(dbToken)); err != nil {
+			return nil, fmt.Errorf("verifying bucket update: %v", err)
+		}
+	}
+
+	// if moving root into subdir, rm .textileseed from subdir
+	if cleanPth == "" {
+		filepath := gopath.Join(cleanDst, buckets.SeedName)
+		_, err = s.removePath(ctx, dbID, dbToken, &pb.RemovePathRequest{
+			Key:  req.Key,
+			Path: filepath,
+		}, filepath)
+
+		if err != nil {
+			return nil, fmt.Errorf("error cleaning path: %v", err)
+		}
+	}
+
+	// a nice absolute path to skip Rel errors
+	ipfsPath := gopath.Join("/ipfs", n.Cid().String(), cleanPth)
+	ipfsDest := gopath.Join("/ipfs", n.Cid().String(), cleanDst)
+
+	// If "a/b" => "a/" no cleanup is needed
+	parent, err := isSubPath(ipfsDest, ipfsPath)
+	if err != nil {
+		return nil, fmt.Errorf("path cleanup failed: %v", err)
+	}
+	if parent == true {
+		return &pb.MovePathResponse{}, nil
+	}
+
+	// If "a/" => "a/b" cleanup each leaf in "a" that isn't "b" (skipping .textileseed)
+	subdir, err := isSubPath(ipfsPath, ipfsDest)
+	if err != nil {
+		return nil, fmt.Errorf("path cleanup failed: %v", err)
+	}
+	if subdir == true {
+		item, err := s.ListPath(ctx, &pb.ListPathRequest{
+			Key:  req.Key,
+			Path: cleanPth,
+		})
+		if err != nil {
+			return nil, fmt.Errorf("path cleanup failed: %v", err)
+		}
+		for _, chld := range item.Item.Items {
+			if strings.Compare(chld.Name, buckets.SeedName) == 0 || isEqualPath(chld.Path, ipfsDest) {
+				continue
+			}
+
+			sp := stripNodePath(chld.Path)
+			if cleanDst == sp {
+				continue
+			}
+			spClean := cleanPath(sp)
+
+			_, err = s.removePath(ctx, dbID, dbToken, &pb.RemovePathRequest{
+				Key:  req.Key,
+				Path: spClean,
+			}, spClean)
+			if err != nil {
+				return nil, fmt.Errorf("path cleanup failed: %v", err)
+			}
+		}
+		if err != nil {
+			return nil, fmt.Errorf("path cleanup failed: %v", err)
+		}
+		return &pb.MovePathResponse{}, nil
+	}
+
+	// if a/ => b/ remove a
+	_, err = s.removePath(ctx, dbID, dbToken, &pb.RemovePathRequest{
+		Key:  req.Key,
+		Path: cleanPth,
+	}, cleanPth)
+	if err != nil {
+		return nil, fmt.Errorf("path cleanup failed: %v", err)
+	}
+
+	return &pb.MovePathResponse{}, nil
+}
+
+// isEqualPath returns true if dest equals src
+func isEqualPath(src, dest string) bool {
+	return filepath.Clean(src) == filepath.Clean(dest)
+}
+
+// isSubPath returns true if chld is a subpath of pth
+func isSubPath(src, dest string) (bool, error) {
+	p, err := filepath.Rel(src, dest)
```

Comment:
You should be able to use `path` here since ipfs paths always use "/"

Reply:
expand on that, not sure i follow the suggestion

Reply:
Never mind, `path` doesn't have a rel method. Though I think that's probably because you don't need the complexity of rel here... you can just compare prefixes, i.e., if src is a prefix of dest, you're in a subpath.

Reply:
(`filepath` is for working with files on a filesystem, that can be addressed in complex ways with .., ~, etc. wich vary from os to os)